File: ios/PlaytomicFoundation/model/DateRange.swift
//
//  DateRange.swift
//  PlaytomicFoundation
//
//  Created by Angel Luis Garcia on 15/03/2017.
//  Copyright © 2017 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public class DateRange {
    public let from: Date
    public let to: Date

    public init(from: Date, to: Date) {
        self.from = from
        self.to = to
    }

    public init(at date: Date, spanMinutes: Int) {
        self.from = date.add(minutes: -spanMinutes)
        self.to = date.add(minutes: spanMinutes)
    }

    public var spanMinutes: Int {
        Int(round((to.timeIntervalSinceReferenceDate - from.timeIntervalSinceReferenceDate) / (2.0 * 60.0)))
    }

    public func contains(date: Date) -> Bool {
        from <= date && date <= to
    }

    public func allDates(spanMinutes: Int, includeLast: Bool = false) -> [Date] {
        var dates = [Date]()
        var lastDate = from
        while lastDate < to {
            dates.append(lastDate)
            lastDate = lastDate.add(minutes: spanMinutes)
        }
        if includeLast && lastDate == to {
            dates.append(lastDate)
        }
        return dates
    }

    public func adding(minutes: Int) -> DateRange {
        DateRange(from: from.add(minutes: minutes), to: to.add(minutes: minutes))
    }

    public func adding(days: Int) -> DateRange {
        DateRange(from: from.add(days: days), to: to.add(days: days))
    }
}

public class TimeZoneDateRange: DateRange {
    public let timeZone: TimeZone

    public init(from: Date, to: Date, timeZone: TimeZone = .current) {
        self.timeZone = timeZone
        super.init(from: from, to: to)
    }

    public init(at date: Date, spanMinutes: Int, timeZone: TimeZone = .current) {
        self.timeZone = timeZone
        super.init(at: date, spanMinutes: spanMinutes)
    }

    public convenience init(dateRange: DateRange, timeZone: TimeZone = .current) {
        self.init(from: dateRange.from, to: dateRange.to, timeZone: timeZone)
    }

    public func resetTimeZone(timeZone: TimeZone) -> TimeZoneDateRange {
        guard timeZone != self.timeZone else { return self }
        let dateFrom = from.dateByRemovingTimeZoneOffset(self.timeZone).dateByAddingTimeZoneOffset(timeZone)
        let dateTo = to.dateByRemovingTimeZoneOffset(self.timeZone).dateByAddingTimeZoneOffset(timeZone)
        return TimeZoneDateRange(from: dateFrom, to: dateTo, timeZone: timeZone)
    }
}

extension DateRange: Equatable { }
public func == (lhs: DateRange, rhs: DateRange) -> Bool {
    lhs.from == rhs.from && lhs.to == rhs.to
}

public func == (lhs: TimeZoneDateRange, rhs: TimeZoneDateRange) -> Bool {
    lhs.from == rhs.from && lhs.to == rhs.to && lhs.timeZone == rhs.timeZone
}

public extension DateRange {
    func toTimeRange(timeZone: TimeZone) -> TimeRange {
        TimeRange(from: from.toTime(timeZone: timeZone), to: to.toTime(timeZone: timeZone))
    }

    func openingRangeTo() -> DateRange {
        DateRange(from: from, to: to.add(seconds: -1))
    }
}

extension DateRange: CustomDebugStringConvertible {
    public var debugDescription: String {
        return "\(type(of: self))(from=\(from), to=\(to))"
    }
}

File: android/app/src/main/java/com/playtomic/test/foundation/model/DateRange.kt
package com.playtomic.foundation.model

import com.playtomic.foundation.extension.addDays
import com.playtomic.foundation.extension.addMinutes
import com.playtomic.foundation.extension.addSeconds
import com.playtomic.foundation.extension.dateByAddingTimeZoneOffset
import com.playtomic.foundation.extension.dateByRemovingTimeZoneOffset
import com.playtomic.foundation.extension.toTime
import java.util.Date
import java.util.TimeZone

/**
 * Created by agarcia on 16/03/2017.
 */

open class DateRange {

    val from: Date
    val to: Date

    constructor(from: Date, to: Date) {
        this.from = from
        this.to = to
    }

    constructor(at: Date, spanMinutes: Int) {
        this.from = at.addMinutes(-spanMinutes)
        this.to = at.addMinutes(spanMinutes)
    }

    val spanMinutes: Int
        get() = (Math.round((to.time - from.time) / (2.0 * 60.0 * 1000))).toInt()

    override fun equals(other: Any?): Boolean {
        val dateRange = other as? DateRange ?: return false
        return to == dateRange.to && from == dateRange.from
    }

    override fun hashCode(): Int {
        var result = from.hashCode()
        result = 31 * result + to.hashCode()
        return result
    }

    public fun contains(date: Date): Boolean =
        from <= date && date <= to

    public fun allDates(spanMinutes: Int, includeLast: Boolean = false): List<Date> {
        val dates = mutableListOf<Date>()
        var lastDate = from
        while (lastDate < to) {
            dates.add(lastDate)
            lastDate = lastDate.addMinutes(spanMinutes)
        }
        if (includeLast && lastDate == to) {
            dates.add(lastDate)
        }
        return dates
    }

    fun addingMinutes(minutes: Int): DateRange =
        DateRange(from = from.addMinutes(minutes), to = to.addMinutes(minutes))

    fun addingDays(days: Int): DateRange =
        DateRange(from = from.addDays(days), to = to.addDays(days))
}

class TimeZoneDateRange : DateRange {
    companion object {}

    val timeZone: TimeZone

    constructor(from: Date, to: Date, timeZone: TimeZone = TimeZone.getDefault()) : super(from = from, to = to) {
        this.timeZone = timeZone
    }

    constructor(at: Date, spanMinutes: Int, timeZone: TimeZone = TimeZone.getDefault()) : super(at = at, spanMinutes = spanMinutes) {
        this.timeZone = timeZone
    }

    constructor(dateRange: DateRange, timeZone: TimeZone = TimeZone.getDefault()) : this(from = dateRange.from, to = dateRange.to, timeZone = timeZone)

    fun resetTimeZone(timeZone: TimeZone): TimeZoneDateRange {
        if (this.timeZone == timeZone) {
            return this
        }
        val dateFrom = this.from.dateByRemovingTimeZoneOffset(this.timeZone).dateByAddingTimeZoneOffset(timeZone)
        val dateTo = this.to.dateByRemovingTimeZoneOffset(this.timeZone).dateByAddingTimeZoneOffset(timeZone)
        return TimeZoneDateRange(from = dateFrom, to = dateTo, timeZone = timeZone)
    }

    override fun equals(other: Any?): Boolean {
        val timeZoneDateRange = other as? TimeZoneDateRange ?: return false
        return to == timeZoneDateRange.to && from == timeZoneDateRange.from && timeZone == timeZoneDateRange.timeZone
    }

    override fun hashCode(): Int {
        var result = from.hashCode()
        result = 31 * result + to.hashCode()
        return result
    }
}

fun DateRange.toTimeRange(timeZone: TimeZone) =
    TimeRange(from = from.toTime(timeZone = timeZone), to = to.toTime(timeZone = timeZone))

fun DateRange.openingRangeTo() = DateRange(from = from, to = to.addSeconds(-1))
